File: Julia/zip.jl
function zip(x,y)
    if x == [] || y == [] return [] end
    [(popfirst!(x), popfirst!(y)); zip(x,y)]
end

function main()
    show(zip([1, 2], ['a', 'b']))
    println()
    show(zip([1], ['a', 'b']))
end

main()

File: Haskell/zip.hs
import Prelude hiding (zip)

zip :: [a] -> [b] -> [(a,b)]
zip []     ys    = []
zip xs    []     = []
zip (x:xs) (y:ys) = (x,y) : zip xs ys

main :: IO()
main = do print(zip [1, 2] ['a', 'b'])
          print(zip [1] ['a', 'b'])
